### DISCUSSION TITLE
Access denied for user 'mariadb'@'172.17.0.1' (using password: YES)

### DISCUSSION THREAD

**Discussion by ITaluone** [testcontainers/testcontainers-dotnet#1120] [Q&A] [answered]
Hi

I updated from a fairly old release (2.1.0) to latest (3.7.0).
I cleaned up all compile errors like `TestcontainersBuilder<MySqlTestcontainer>` => `new MariaDbBuilder()` and so on.

After I had this achieved I started up all my integration tests, but all failed with the error above. I cannot imagine why this happens. I even tried with the example from documentation which is: `var database = new MariaDbBuilder().Build();`, but even with that i doesn't work.

I running this with docker desktop on windows 10.

Thanks for any advice

**Comment by HofmeisterAn** [comment, marked as answer]
How do you connect to MariaDB? Can you please share how you create the connection string? Can you please try the connection string from the created container instance.

**Reply by ITaluone** [reply]
I use `WebApplicationFactory<T>`.

```csharp
public class DefaultFactory<TProgram, TDbContext> : WebApplicationFactory<TProgram>, IAsyncLifetime
    where TProgram : class
    where TDbContext : DbContext
{
    private readonly MariaDbContainer _database;

    public DefaultFactory()
    {
        _database = new MariaDbBuilder().Build();
    }

    public async Task InitializeAsync() => await _database.StartAsync();

    public new async Task DisposeAsync() => await _database.DisposeAsync();

    protected override void ConfigureWebHost(IWebHostBuilder builder) =>
        builder.ConfigureServices(services =>
        {
            services.RemoveDbContext<TDbContext>();

            services.AddDbContext<TDbContext>(opt =>
            {
                opt.UseMySql(_database.GetConnectionString(),
                    ServerVersion.Parse("10.9.2-mariadb"));
            });

            services.EnsureDbCreated<TDbContext>();
        });
}
```

And then I inject this class into the integration test classes like:

```csharp
public class IntegrationTests : IClassFixture<DefaultFactory<Program, DbContext>>
{
    private readonly DefaultFactory<Program, DbContext> _factory;

    public IntegrationTests(DefaultFactory<Program, DbContext> factory)
    {
        _factory = factory;
    }
}
```

But I don't make it to execute the tests. The error throws inside the `EnsureDbCreated()` extension method, which looks like this:

```csharp
 public static void EnsureDbCreated<TDbContext>(this IServiceCollection services)
        where TDbContext : DbContext
    {
        var serviceProvider = services.BuildServiceProvider();

        using var scope = serviceProvider.CreateScope();
        var scopedServices = scope.ServiceProvider;
        var context = scopedServices.GetRequiredService<TDbContext>();
        context.Database.EnsureCreated(); // <--- this throws
    }
```

**Reply by HofmeisterAn** [reply]
I tested the MariaDB module along with our WeatherForecast example, which also utilizes EF (`Pomelo.EntityFrameworkCore.MySql`), and I had no issues. The WebApp starts successfully, and the tests that use `WebApplicationFactory<TEntryPoint>` also run without any problems. Usually, I do not replace services; instead, I pass the [connection string](https://dotnet.testcontainers.org/examples/aspnet/#utilizing-webapplicationfactory) to the application, which is IMO the better approach. Is it possible that you need to remove other services too (the old [guide](https://testcontainers.com/guides/testing-an-aspnet-core-web-app/) removes `DbContextOptions` and `DbConnection`), and that there are still be some wrong services registered from the `Program` configuration?

**Reply by ITaluone** [reply]
Hmm.. doesn't work, unfortunately.. 

Yeah.. Inside `RemoveDbContext()` I am doing the same like in your guide..

What I am wondering is: the error tells me about an internal docker network ip address (btw.. this is the gateway address), while the connection string shows the correct one (localhost). This is usually a db configuration "error"..
Is it somehow possible to change this configuration?

**Reply by HofmeisterAn** [reply]
> the error tells me about an internal docker network ip address (btw.. this is the gateway address), while the connection string shows the correct one (localhost).

Yes, that is why I asked how you create the connection string. TC will pick the right host. In your case, it should be `127.0.0.1`. TC returns the [gateway](https://github.com/testcontainers/testcontainers-dotnet/blob/5600106cd9ef901a67e2ab3245bcbb21ca64dc8b/src/Testcontainers/Containers/DockerContainer.cs#L149-L150) if you are running inside another container, but I assume your configuration does not, since you mentioned Docker Desktop.

> Is it somehow possible to change this configuration?

Which configuration do you want to change? You can use the build's APIs, the environment variables the Docker image provides, or you can also use the generic container builder to set up your own configuration.

Have you tried to establish a plain database connection and run a simple operation?

**Reply by ITaluone** [reply]
What do you mean with "plain db connection"? I can connect e.g. with HeidiSQL (or any other SQL client)

**Reply by ITaluone** [reply]
Hmm.. I cannot tell why it is working now..

Just to recap what I have changed compared to yesterday:

- I realized, that mariaDB:11 (or latest) have some issues to spin up correctly (meaning the DB containers started, but then nothing more happens (the tests were frozen in "Pending" state) --> so I switched back to mariaDB:10
- I had indeed an error in my connection string: I append some other db options e.g. treattinyintasboolean and similar. Since `_database.GetConnectionString()` does not end with a `;` anymore (I'm sure it appended it in 2.1.0**, because it worked back then by just concatenating the two token)* I changed that
- I now switched to `UseMysql(conn, SemVer.AutoDetect(conn))`, previously I had hardcoded the mariadb version

So... long story short: I cannot tell why it is working (again), possibly a combination of all three.. I don't know.. 
But I am "green" everywhere now, so sorry to bother you and thanks for your time and help :)

\* interestingly the example above didn't work either 🤷‍♂️ 
\** Yepp.. it does ;)
https://github.com/testcontainers/testcontainers-dotnet/blob/53d37fa25ca6523e4f0fc47feee1c5f5f739715c/src/Testcontainers/Containers/Modules/Databases/MySqlTestcontainer.cs#L25-L26C116
